src/drivers/intel/cavs/timer.c

```diff
@@ -36,22 +36,55 @@ void platform_timer_stop(struct timer *timer)
 		   shim_read(SHIM_DSPWCTCS) & ~SHIM_DSPWCTCS_T0A);
 }
 
+/* clock ticks it takes to set clock and enable IRQ */
+#define TIMER_OVERHEAD	500
+
+/* max time it should take to do 2 subsequenmt timer reads */
+#define TIMER_READ_TIME	256
+
+static spinlock_t timer_lock = {0};
+
 int64_t platform_timer_set(struct timer *timer, uint64_t ticks)
 {
+	uint64_t ticks_now[2];
+	uint32_t flags;
+
 	/* a tick value of 0 will not generate an IRQ */
 	if (ticks == 0)
 		ticks = 1;
 
-	/* Check if requested time is not past time */
-	if (ticks > shim_read64(SHIM_DSPWC))
+	spin_lock_irq(&timer_lock, flags);
+
+	/* 64bit reads are non atomic on xtensa so we must
+	 * read a stable value where there is no bit 32 flipping.
+	 * A large delta between reads[0..1] means we have flipped
+	 * and that the value read back in either 0..1] is invalid.
+	 */
+	do {
+		ticks_now[0] = shim_read64(SHIM_DSPWC);
+		ticks_now[1] = shim_read64(SHIM_DSPWC);
+
+		/* worst case is we perform this twice so 4 clock reads */
+	} while (ticks_now[1] - ticks_now[0] > TIMER_READ_TIME);
+
+	/* Check if requested time is not past time and include the
+	 * overhead of changing the timer
+	 */
+	if (ticks > ticks_now[1] + TIMER_OVERHEAD) {
```

Comment:
No need to define separate const for timer overhead, the TIMER_MIN_RECOVER_CYCLES should be used instead. Overall the TIMER_MIN_RECOVER_CYCLES is the safe margin that is used to schedule from 'now' in case of the delay.

Reply:
My thinking here is that this is relative to the DSP clock and not the wallclock. It was a safe guess that may need refinement too, depending on HPCRO, LPCRO etc.
One other thing is that this PR fails on the multcore multkernel tests, but there is nothing obvious in the CI logs ?

Reply:
PM core enable IPCs are failing, they're timing out.

Reply:
@slawblauciak thanks - btw we have
```c
	while (!cond(target_core)) {
		if (deadline < platform_timer_get(timer)) {
			/* safe check in case we've got preempted
			 * after read
			 */
			if (cond(target_core))
				break;

			tr_err(&idc_tr, "idc_wait_in_blocking_mode() error: timeout");
			return -ETIME;
		}
	}
```
So I'm assuming here we are slowing boot down in the other core (since we are spinning reading the time). I'll send an update.